packages/babel-parser/src/parser/error.js

```diff
@@ -64,31 +72,26 @@ export {
   SourceTypeModuleErrorMessages as SourceTypeModuleErrors,
 } from "./error-message";
 
-export type raiseFunction = (number, ErrorTemplate, ...any) => void;
+export type raiseFunction = (ErrorTemplate, Origin, ...any) => void;
+export type ErrorData = {| message: ErrorTemplate, loc: Position |};
 
 export default class ParserError extends CommentsParser {
   // Forward-declaration: defined in tokenizer/index.js
   /*::
   +isLookahead: boolean;
   */
 
-  getLocationForPosition(pos: number): Position {
-    let loc;
-    if (pos === this.state.start) loc = this.state.startLoc;
-    else if (pos === this.state.lastTokStart) loc = this.state.lastTokStartLoc;
-    else if (pos === this.state.end) loc = this.state.endLoc;
-    else if (pos === this.state.lastTokEnd) loc = this.state.lastTokEndLoc;
-    else loc = getLineInfo(this.input, pos);
-
-    return loc;
-  }
-
   raise(
-    pos: number,
     { code, reasonCode, template }: ErrorTemplate,
+    origin: Origin,
     ...params: any
   ): Error | empty {
```

Comment:
I actually originally had this, but decided against it for a couple reasons:

1. I wanted it to read like throwing an error "throw SomeError(metadata)", which calls out what matters most at that line "hey, this is a MissingSemicolonError" or "hey, this is RestTrailingCommaError"  vs. front-loading the information that matters least, the associated metadata.  when reading the code, all the metadata, like doing throw "metadata on SomeError";
2. I think we will want potentially want to add more to this object later. As I mentioned in the above, we may eventually want to expand this to `{ at: length: }` for ranges (or `{ start, end }` or however you want to do that), and I think it would also eventually be nice to be able to supply context info (we currently have this info but it gets stringified for example, but can be very useful to the user to be able to do error.enumName vs. parsing message for enumName). As such, I wanted to put in the scaffolding for something that could carry additional info that could be easily read in the future. 
3. From my testing there seems to be zero performance benefit to avoiding object creation at this point, for two reasons. First, it is a seldom occurring event, as it happens once per error raise (vs. getLineInfo which was an iterative event that can lead to an O(N) slowdown per Error, where N = source length). Secondly, if we did consider it to be performance sensitive, then we'd for example be much better served by instead changing `raiseWithData` to take (missingPlugin?: Array<string>, code?: string)` instead of `{ missingPlugin?: Array<string>,  code?: string, }` (since this object `data` object gets thrown away immediately when yet another object is created on the corresponding call to `_raise`), and this change would basically affect *all* error raises "for free" transparently since it's an internal function call in `raise`. However, I also don't think this is a performance issue.

Reply:
Ok, sounds good :+1: